src/daemon/led_wireless.c

```diff
@@ -1,62 +1,171 @@
+    HWANIM_TYPELIGHTNING_RIPPLE = 0x09,
+    HWANIM_NONE                 = 0xff
+};
 
-    // Dark Core commands:
-    // 00: Colour Shift
-    // 01: Colour Pulse
-    // 03: Rainbow
-    // 07: Static Colour
-    // FF: No animation (black)
+enum HWANIM_ZONE_BIT {
+    HWANIM_ZONEBIT_FRONT = 0,
+    HWANIM_ZONEBIT_THUMB = 1,
+    HWANIM_ZONEBIT_REAR  = 2
+};
 
+enum HWANIM_DIRECTION {
+    HWANIM_DIRECTION_NONE  = 0,
+    HWANIM_DIRECTION_LEFT  = 1,
+    HWANIM_DIRECTION_RIGHT = 2
+};
 
+enum HWANIM_SPEED {
+    HWANIM_SPEED_NONE   = 0,
+    HWANIM_SPEED_LOW    = 1,
+    HWANIM_SPEED_MEDIUM = 2,
+    HWANIM_SPEED_HIGH   = 3
+};
+
+int updatergb_wireless(usbdevice* kb, lighting* lastlight, lighting* newlight) {
     if(IS_DARK_CORE(kb)) {
-        for(int zone = 0; zone < 3; zone++) {
-            uchar data_pkt[MSG_SIZE] = {
-                CMD_SET, 0xaa, 0
-            };
-
-            data_pkt[4]  = 1 << zone;        // Bitmask.
-            data_pkt[5]  = 7;                // Command (static colour).
-            data_pkt[8]  = 100;              // Opacity (100%).
-            data_pkt[9]  = newlight->r[zone];
-            data_pkt[10] = newlight->g[zone];
-            data_pkt[11] = newlight->b[zone];
-            // Colour gets sent twice for some reason.
-            data_pkt[12] = newlight->r[zone];
-            data_pkt[13] = newlight->g[zone];
-            data_pkt[14] = newlight->b[zone];
-            data_pkt[15] = 5 - zone;         // Profile byte?
-
-            if(!usbsend(kb, data_pkt, 1))
-                return -1;
+        for(int i = 0; i < 3; i++) {
+            char code[20] = "";
+            snprintf(code, sizeof(code), "7:1:0:%02hhx%02hhx%02hhx", newlight->r[i], newlight->g[i], newlight->b[i]);
+            cmd_hwanim(kb, 0, 0, 190 + i, code);
         }
-    } else if (IS_K63_WL(kb)) {
-        uchar data_pkt[MSG_SIZE] = {
-            CMD_SET, 0xaa, 0
-        };
-
-        data_pkt[4]  = 0;
-        data_pkt[5]  = 7;
-        data_pkt[8]  = 100;
-        data_pkt[9]  = newlight->r[0];
-        data_pkt[10] = newlight->g[0];
-        data_pkt[11] = newlight->b[0];
-        // Colour gets sent twice for some reason.
-        data_pkt[12] = newlight->r[0];
-        data_pkt[13] = newlight->g[0];
-        data_pkt[14] = newlight->b[0];
-        data_pkt[15] = 0;
-
-        if(!usbsend(kb, data_pkt, 1))
-            return -1;
     }
 
     memcpy(lastlight, newlight, sizeof(lighting));
     return 0;
 }
+
+void cmd_hwanim(usbdevice* kb, usbmode* dummy1, int dummy2, int keyindex, const char* code) {
+    // Unified format for the code string:
+    //  - DARK CORE Format: animation mode:speed:random:colour:colour2
+    //  - K63 WL    Format: animation mode:speed:dir
+    
+    uchar anim = 0xff, speed = 2, rand_or_dir = 0;
+    uchar r[2] = {}, g[2] = {}, b[2] = {};
+    sscanf(code, "%hhx:%hhu:%hhu:%02hhx%02hhx%02hhx:%02hhx%02hhx%02hhx", &anim, &speed, &rand_or_dir, r, g, b, r+1, g+1, b+1);
```

Comment:
While we certainly could do that, it would make it impossible to just leave out parameters you don't need for a certain animation mode. For example, `hwanim zone1:3:1` sets the animation to a rainbow at slow speed on zone 1. We don't need to pass a random/direction value or a color here. The variables are all initialized for this exact reason.

Reply:
Makes sense. As long as it doesn't send packets/change state by feeding it garbage, it should be okay.